<h3>Задача семинара</h3>

<p>Мы будем решать задачу, о которой говорили на лекции &mdash;&nbsp;реализуем&nbsp;механизм подсказок (<em>suggest</em>) категории по заголовку.</p>

<p>Представьте, что вы хотите разместить объявление о продаже шин. Чтобы положить своё объявление в нужную категорию, вам необходимо выбрать одну из 10 верхнеуровневых категорий, одну из 5 на следющем уровне, одну из 12 на следующем уровне, одну из 5 на последнем уровне.</p>

<p>Даже если варианты правильных&nbsp;категорий&nbsp;на каждом уровне очевидны, процесс подачи объявления лучше упростить.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="259" src="http://labelimages.avito.ru/mlcourse_week3_without_suggest.png" width="563" />
<figcaption>Пример поиска нужной категории без подсказок</figcaption>
</figure>
</div>

<p>&nbsp;Упрощённый процесс подачи объявлений может выглядеть следующим образом. Пользователь заполняет заголовок объявления, затем выбирает один из предложенных вариантов категорий, который больше всего подходит его объявлению.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="324" src="http://labelimages.avito.ru/mlcourse_week3_with_suggest.png" width="534" />
<figcaption>Работа suggest при публикации объявления</figcaption>
</figure>
</div>

<p>Задача &mdash; определить категорию объявления, используя заголовок. Всего в выборке будет 54 категории.</p>


<h3><strong>Многоклассовая классификация</strong></h3>

<p>Для решения задачи будем использовать модель логистической регрессии – линейная модель для бинарной классификации. Как её использовать для решения задачи классификации на 54 класса? Оказывается, любой бинарный классификатор, который способен оценивать меру принадлежности объектов к одному из 2 классов, можно использовать для многоклассовой классификации. </p>

<p>Мнгоклассовая классификация на основе бинарного классификатора может быть реализована с помощью одной из двух стратегий:</p>

<ul>
	<li>Один против всех (<em>One-vs-Rest</em> или <em>One-vs-All</em>);</li>
	<li>One-vs-One.</li>
</ul>

<p> </p>

<p><strong>One-vs-All </strong></p>

<p>В стратегии <em>One-vs-All</em> строятся <span class="math-tex">\(k\)</span> классификаторов, отделяющих каждый класс от остальных. Например, для 3 классов нужно обучить 3 классификатора:</p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="436" name="Снимок экрана 2021-08-03 в 19.26.56.png" src="https://ucarecdn.com/2abe5812-3f6e-42cf-97a4-a35a6de58c6d/" width="513">
<figcaption>Многоклассовая классфикация по стратегии OVR</figcaption>
</figure>
</div>

<p>Помимо эффективности построения (необходимы только <span class="math-tex">\(k\)</span> классификаторов), одним из преимуществ этого подхода является его интерпретируемость. Поскольку для каждого класса обучался свой классификатор, можно анализировать каждый классификатор в отдельности, чтобы понять, почему модель приняла такое решение.</p>

<p><em>Предсказания на тестовой выборке для OVR.</em></p>

<p>Для объектов из тестовой выборки предсказывается класс, получивший максимальную вероятность в результате предсказаний <span class="math-tex">\(k\)</span> классификаторов:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \ a_{k}(x),\)</span></p>

<p>где <span class="math-tex">\(a_k(x) -\)</span> предсказание модели бинарной классификации для класса <span class="math-tex">\(k\)</span>.  Выражение<em> <em><em><em><em><em><em><em><span class="math-tex">\(\argmax\)</span></em></em></em></em></em></em></em></em> означает выбор <span class="math-tex">\(k\)</span>, при котором значение <span class="math-tex">\(a_{k}(x)\)</span> максимально.</p>

<p><em>На практике обычно используют OVR стратегию или алгоритмы с нативной поддержкой многоклассовой классификации – нейронные сети или алгоритмы, которые мы будем изучать дальше в этом курсе (RandomForest, KNN).</em></p>

<p> </p>

<p><strong>One-vs-One (опционально)</strong></p>

<p>В стратегии <em>One-vs-One </em>обучается один классификатор для каждой пары классов. При этом каждый классификатор обучается только на объектах, относящихся к этим классам: </p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="551" name="Снимок экрана 2021-08-01 в 23.43.15.png" src="https://ucarecdn.com/b6a950d0-a787-49de-b6cd-9f2faadb91c3/" width="637">
<figcaption>Многклассовая классификация стратегией OVO</figcaption>
</figure>
</div>

<p>Таким образом, требуется построить квадратичное число классификаторов <span class="math-tex">\(\frac{k(k-1)}{2}\)</span>.</p>

<p>При больших <span class="math-tex">\(k\)</span> может потребоваться много времени на обучение. Однако этот метод может быть полезен для алгоритмов, которые плохо масштабируются с увеличением размера выборки, так как каждая отдельная задача обучения классификатора включает в себя только объекты, принадлежащие двум выбранным классам. </p>

<p><em>Предсказания на тестовой выборке для OVO.</em></p>

<p>Для объектов из тестовой выборки все <span class="math-tex">\(\frac{k(k-1)}{2}\)</span> классификаторов делают предсказания, а финальный ответ получается в результате «жёсткого» голосования:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \sum_{m \neq k} [a_{km}(x) = 1],\)</span></p>

<p>где  <span class="math-tex">\(a_{km}(x) -\)</span>классификатор с позитивным классом <span class="math-tex">\(k\)</span> и негативным классом <span class="math-tex">\(m\)</span>.  Жёстким оно называется потому, что никак не учитывается, насколько уверены были классификаторы в своих предсказаниях.</p>

<p>Если несколько классов для объекта получили одинаковое количество голосов или нужно определить не только класс, но и меру принадлежности, производится «мягкое» голосование: выигрывает тот класс в котором суммарно больше уверены классификаторы.</p>


## EDA

<p>Начнем работу с данными с импорта нужных библиотек:</p>

In [2]:
from google.colab import drive
drive.mount('/content/drive')
way = "/content/drive/MyDrive/Colab_Notebooks/aaa_ml/datasets/"

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

<p>Прочитаем <a href="https://stepik.org/media/attachments/lesson/537384/suggest_train.csv" rel="noopener noreferrer nofollow">данные</a>:</p>


In [4]:
data = pd.read_csv(way+'suggest_train.csv', index_col=0)
data.head(3)

title  category_id  \
item_id                                                                   
335330            Раствор. Товарный бетон от производителя           15   
137480                                    Детская кроватка           16   
242690   Кровать -качалка с ящиком для вещей, два полож...           34   

                                                      name  
item_id                                                     
335330   Для дома и дачи|Ремонт и строительство|Стройма...  
137480   Для дома и дачи|Мебель и интерьер|Кровати, див...  
242690   Личные вещи|Товары для детей и игрушки|Детская...

Видим, что данные упорядочены по категориям. Это может привести к ошибке при разбиении данных. Перемешаем данные:

In [5]:
data = data.sample(frac=1).reset_index(drop=True)

Проверим, правильно ли закодированы категории:

In [6]:
np.sort(data.category_id.unique())

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53])

In [7]:
data.category_id.min(), data.category_id.max()

(0, 53)

Категории закодированы правильно.

**Баланс классов**

Посмотрим на распределение классов. Большой дисбаланс может вызвать проблемы модели. Например, в топ-3 по вероятности будут попадать нерелевантные категории просто потому, что в датасете их намного больше, чем остальных.

In [8]:
data.groupby('category_id').size().describe()

count     54.000000
mean     370.370370
std       33.178361
min      309.000000
25%      346.000000
50%      361.000000
75%      398.750000
max      447.000000
dtype: float64

Видим, что минимальное количество примеров на категорию 309, а максимальное 447. Скорее всего, проблемы не возникнет.

**Примеры заголовков**

Посмотрим на примеры топ-3 наиболее и наименее популярных категорий:

In [9]:
data[data.name.isin(data.name.value_counts()[:3].index)].drop_duplicates(subset='name')

title  category_id  \
58                   Nuga Best           37   
70  Ковёр напольный 200х300 см           27   
81            Зеркало для дома           23   

                                                 name  
58  Личные вещи|Красота и здоровье|Приборы и аксес...  
70  Для дома и дачи|Мебель и интерьер|Текстиль и к...  
81           Для дома и дачи|Мебель и интерьер|Другое

Наименее популярные:

In [10]:
data[data.name.isin(data.name.value_counts()[-3:].index)].drop_duplicates(subset='name')

title  category_id  \
18     Продам ноутбук            1   
23  Телевизор thomson            5   
63     Sokolov кольцо           40   

                                                 name  
18                       Бытовая электроника|Ноутбуки  
23  Бытовая электроника|Аудио и видео|Телевизоры и...  
63     Личные вещи|Часы и украшения|Ювелирные изделия

Видим, что в бытовой технике могут быть довольно редкие сочетания букв, например "K50AB", есть слова с орфографическими ошибками ("гаршок"). Есть и частые слова, которые встретились в 2 их 6 объявлений.

**Названия категорий**

Посмотрим на названия категорий:

In [11]:
data.name.unique()[:3]

array(['Для дома и дачи|Ремонт и строительство|Стройматериалы',
       'Личные вещи|Одежда, обувь, аксессуары|Аксессуары',
       'Бытовая электроника|Телефоны|Samsung'], dtype=object)

Видим, что категории могут находиться на разном уровне иерархии. Посмотрим, насколько сильно может отличаться глубина категории:

In [12]:
data.name.apply(lambda x: len(x.split('|'))).value_counts().sort_index()

2     1406
3    15973
4     2621
Name: name, dtype: int64

Можно подумать, как использовать данные о том, что у классов есть иерархическая структура.

**Статистики по текстам**

Посмотрим, отличаются ли статистически тексты заголовков для разных категорий. Посчитаем количество уникальных символов в заголовке:

In [13]:
data.title.apply(len).describe()

count    20000.000000
mean        25.093300
std         12.474148
min          5.000000
25%         15.000000
50%         23.000000
75%         34.000000
max         50.000000
Name: title, dtype: float64

Посмотрим на на распределение количества слов:

In [14]:
data.title.apply(lambda x: len(x.split())).describe()

count    20000.00000
mean         3.79825
std          1.96368
min          1.00000
25%          2.00000
50%          3.00000
75%          5.00000
max         14.00000
Name: title, dtype: float64

В среднем в заголовках 3-4 слова. Заголовки короткие

Интересно посмотреть на краевые случаи. Давайте выведем примеры, где в заголовок получилось уместить больше 12 слов:

In [15]:
data[data.title.apply(lambda x: len(x.split()) > 12)].head(3)

title  category_id  \
7772     Отдам 2 альбома для фото 10 х 15 на 96 и 48 фот           19   
8215    iPad pro 128 4G + клва + стилус + сумка и так. Д            3   
16502  PS3 Slim + Ps Move + Ps Eye + 3 джойстика + 12...            4   

                                                    name  
7772   Для дома и дачи|Мебель и интерьер|Предметы инт...  
8215   Бытовая электроника|Планшеты и электронные кни...  
16502  Бытовая электроника|Игры, приставки и программ...

Все примеры из бытовой электроники. Значит, количество слов может послужить предиктором категории. Посчитаем статистики для категорий:

In [16]:
data['word_count'] = data.title.apply(lambda x: len(x.split()))
data.groupby('name')['word_count'].mean().sort_values().head(3)

name
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Платья и юбки     2.576190
Для дома и дачи|Мебель и интерьер|Шкафы и комоды                       2.756219
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Верхняя одежда    2.824074
Name: word_count, dtype: float64

Действительно, среднее количество слов для категории отличается в 2 раза.

In [17]:
data.groupby('name')['word_count'].mean().sort_values().tail(3)

name
Хобби и отдых|Велосипеды|Запчасти и аксессуары            5.508600
Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки    5.531429
Хобби и отдых|Билеты и путешествия|Путешествия            5.641399
Name: word_count, dtype: float64

Кажется, что статистики по тексту могут помочь улучшить качество работы классификатора. С другой стороны, нужно хорошенько подумать, стоит ли использовать такие признаки в сервисе?

In [34]:
data.shape

(20000, 4)

In [35]:
data_fixed = data.drop_duplicates(subset='title')

In [36]:
data_fixed.shape

(17925, 4)

## Бейзлайн

Разобьем данные на обучающую выборку и валидацию:

In [18]:
train_titles, val_titles, y_train, y_val = train_test_split(
    data['title'],
    data['category_id'],
    random_state=1, stratify=data['category_id']
)

Обычно, в задачах классификации первым бейзлайном является предсказание самым популярным классом. Но когда выборка сбалансирована, а классов много, такой бейзлайн не очень информативен:

In [19]:
(y_val == y_train.mode().values[0]).mean()

0.0224

## CountVectorizer

Для получения матриц признаков преобразуем заголовки с помощью CountVectorizer():

In [20]:
count_vectorizer = CountVectorizer()

X_train = count_vectorizer.fit_transform(train_titles.values)
X_val = count_vectorizer.transform(val_titles.values)

In [21]:
X_train

<15000x17211 sparse matrix of type '<class 'numpy.int64'>'
	with 54694 stored elements in Compressed Sparse Row format>

<p>В данном случае вместо&nbsp;<span class="math-tex">\(15000 \times 17316=259740000\)</span>&nbsp;значений мы храним всего&nbsp;<span class="math-tex">\(54676\)</span> тройки&nbsp;<span class="math-tex">\((i, j, x_j^{(i)})\)</span>. Это&nbsp;1500&nbsp;раз меньше!</p>

<p>Sparse формат не доставляет никаких неудобств &mdash;&nbsp;все полезные функции можно применять к матрицам в разреженном формате. Единственное &laquo;неудобство&raquo;&nbsp;&mdash;&nbsp;нельзя посмотреть&nbsp;на матрицу.<span style="color:#a03881"> </span>Но в данном случае этого делать не нужно, потому что всё, что мы увидим, это матрица&nbsp;из нулей. Не повторяйте этого на практике, на больших&nbsp;выборках преобразованные данные могут не поместиться в оперативную память:</p>

In [22]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

Посмотрим на получившийся словарь:

In [23]:
print('words in vocab: ', len(count_vectorizer.vocabulary_))
count_vectorizer.vocabulary_

words in vocab:  17211


{'кожаный': 10566,
 'чехол': 16648,
 'индийские': 10058,
 'вазы': 8256,
 'летняя': 11278,
 'коляска': 10633,
 'шкаф': 16813,
 'угловой': 15913,
 'комплект': 10684,
 'бутылочек': 8220,
 'шорты': 16861,
 'лавка': 11130,
 'со': 14969,
 'спинкой': 15109,
 'кровать': 10978,
 'детская': 9299,
 'отражательная': 12822,
 'печь': 13131,
 'телевизор': 15561,
 'люстра': 11452,
 'dia750': 3153,
 'tt40': 6666,
 'wg': 6993,
 'wi': 7004,
 'fi': 3619,
 'модуль': 11936,
 'продам': 13779,
 'пуховое': 13934,
 'одеяло': 12622,
 'часы': 16586,
 'rado': 5714,
 'продаю': 13781,
 'acer': 2087,
 'aspire': 2284,
 'v3': 6771,
 'нетбук': 12384,
 'отличном': 12814,
 'состоянии': 15065,
 'английский': 7536,
 'для': 9417,
 'бакалавров': 7714,
 'антиквариат': 7565,
 'рога': 14249,
 'лося': 11391,
 'приблиз': 13692,
 '1846': 492,
 'табурет': 15461,
 '12': 207,
 '18': 466,
 'wifi': 7008,
 'роутер': 14333,
 'новый': 12446,
 'дом': 9453,
 'ру': 14339,
 'i7': 4175,
 '16gb': 415,
 'ram': 5724,
 'монитор': 11992,
 'viewsonic

Убедимся, что необходимые значения хранятся по индексам из словаря:

In [24]:
count_vectorizer.transform(['Продам продам'])[0, count_vectorizer.vocabulary_['продам']]

2

Видим, что CountVectorizer автоматически приводит слова к нижнему регистру.

Важные параметры

Класс CountVectorizer содержит все нужные параметры для настройки преобразования под вашу задачу:

<pre>
<code class="language-python">CountVectorizer(
    *,
    input='content',
    encoding='utf-8',
    decode_error='strict',
    strip_accents=None,
    lowercase=True,
    preprocessor=None,
    tokenizer=None,
    stop_words=None,
    token_pattern='(?u)\\b\\w\\w+\\b',
    ngram_range=(1, 1),
    analyzer='word',
    max_df=1.0,
    min_df=1,
    max_features=None,
    vocabulary=None,
    binary=False,
    dtype=&lt;class 'numpy.int64'&gt;,
)</code></pre>

<p>Рассмотрим наиболее важные из них:</p>

<ul>
	<li><code>max_df</code>, <code>min_df</code>&nbsp;&mdash;&nbsp;параметры, позволяющие фильтровать слова по порогу встречаемости. Если вы передаёте в параметры действительные числа от 0 до 1, это это максимальная и минимальная доля встречаемости слова. Если целые, то абсолютное значение встречаемости;</li>
	<li><code>ngram_range</code>&nbsp;&mdash;&nbsp;диапазон n-грамм: (1, 3) &mdash;&nbsp;значит, включить в словарь&nbsp;униграммы, биграммы и тригаммы;</li>
	<li><code>analyzer</code>&nbsp;&mdash;&nbsp;определяет, что считать токеном. Есть возможность токенезировать по словам, буквам и буквам с учётом окончаний слов. Может работать особенно эффективно в сочетании с учётом n-грамм;</li>
	<li><code>token_pattern</code>&nbsp;&mdash; регулярное выражение для токенизации.</li>
</ul>


## Модель логистической регрессии

Обучим модель логистической регрессии c параметрами по умолчанию и посчитаем accuracy:

In [25]:
model = LogisticRegression()
model.fit(X_train, y_train)

preds = model.predict(X_val)
accuracy_score(preds, y_val)

0.7532

Логистическая регрессия позволяет получить не только предсказанные категории, но и вероятности. Предсказания для одного объекта в нашем случае выглядят как вектор из 54 чисел:

In [30]:
np.round(model.predict_proba(X_val[0]), 3)

array([[0.003, 0.003, 0.004, 0.74 , 0.004, 0.002, 0.011, 0.005, 0.005,
        0.001, 0.005, 0.003, 0.005, 0.002, 0.004, 0.006, 0.005, 0.004,
        0.007, 0.003, 0.004, 0.009, 0.005, 0.009, 0.005, 0.003, 0.006,
        0.006, 0.005, 0.009, 0.004, 0.008, 0.005, 0.008, 0.006, 0.003,
        0.002, 0.007, 0.011, 0.005, 0.003, 0.004, 0.002, 0.002, 0.004,
        0.003, 0.006, 0.005, 0.001, 0.009, 0.008, 0.002, 0.004, 0.002]])

Проверим свою интуицию по поводу весов полученной модели. Мы знаем, что слова встречающиеся в разных категориях, не должны получить большой вес в модели. Посмотрим на вес слова 'продам' для разных категорий:

In [27]:
model.coef_[:, count_vectorizer.vocabulary_['продам']]

array([ 0.61508321,  0.0275319 ,  0.23054164,  0.19648764,  0.31772916,
       -0.41027092,  0.25355874,  0.53986652,  0.16609963, -0.87113178,
        0.56714439,  0.43577777, -0.15120484, -0.16935711,  0.16280601,
       -0.34820924,  0.61205611, -0.66032679,  0.67725215, -0.6668873 ,
       -0.61100196,  0.36765554,  0.51399808,  0.18882214, -0.4328369 ,
        0.34304989,  0.63465919,  0.15821976,  0.19361738,  0.5392335 ,
       -0.47698731,  0.37718243,  0.41415103,  0.9015056 ,  0.48126266,
        0.30593524, -0.55769946,  0.04148851,  0.60525324,  0.20016629,
       -0.04959199,  0.09357258, -0.43902841, -1.21673181, -0.81983805,
       -0.3568331 ,  0.12378717, -0.40034697, -1.313023  ,  0.27784154,
        0.61348661, -1.19245123, -0.32405477, -0.70901031])

Слова, по которым можно определить категорию, наоборот, должны получить большой вес. Посмотрим на вес для слова 'iphone' для разных категорий:

In [28]:
model.coef_[:, count_vectorizer.vocabulary_['iphone']]

array([ 6.21963244,  0.24719949, -0.18563037, -0.36220082, -0.17216695,
       -0.17364277,  1.04034185, -0.17964915, -0.22922341,  3.42803625,
        0.5893871 , -0.11887824,  0.29830444, -0.1650981 , -0.28042387,
       -0.3949644 , -0.15845648, -0.34888094, -0.20959327, -0.31174124,
       -0.3218078 , -0.31550465, -0.18074611, -0.38632137, -0.35991005,
       -0.13071649, -0.19006903, -0.30286555, -0.21860909, -0.27587543,
       -0.34448624, -0.31359404, -0.18982853, -0.18939669, -0.21535792,
       -0.12281622, -0.18494629, -0.39138513, -0.31275579, -0.21191184,
       -0.29793342, -0.24450839, -0.1848781 , -0.3366014 , -0.35378464,
       -0.22157762,  0.30129324, -0.41215437, -0.1827871 , -0.35802705,
       -0.21748632, -0.26653211, -0.29096919, -0.30750082])

Действительно, вес при этом слове для некоторых категорий очень большой. Ведь если это слово есть в заголовке, скорее всего, правильной категорией будет что-то начинающееся на 'Бытовая электроника|Телефоны|'. Проверим, так ли это:

In [29]:
cat_names = data.set_index('category_id')['name'].to_dict()
top_iphone_cats = np.argsort(model.coef_[:, count_vectorizer.vocabulary_['iphone']])[-3:]
[cat_names[c] for c in top_iphone_cats]

['Бытовая электроника|Телефоны|Другие марки',
 'Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки',
 'Бытовая электроника|Телефоны|iPhone']

**Ошибки модели**

Посмотрим, в каких категориях чаще всего происходят ошибки. Для этого создадим удобный dataframe:

In [32]:
models_erros = pd.DataFrame({
    'title': val_titles.values,
    'preds_category': [cat_names[cat] for cat in preds],
    'true_category': [cat_names[cat] for cat in y_val],
    'preds_ok': preds == y_val
})

Выведем 3 случайные ошибки модели:

In [33]:
models_erros[~models_erros['preds_ok']].sample(3)

title  \
8770            Спальный гарнитур   
7744  Monster High/Toralei Stripe   
6853             Пледы с рукавами   

                                         preds_category  \
8770  Для дома и дачи|Мебель и интерьер|Кровати, див...   
7744     Личные вещи|Товары для детей и игрушки|Игрушки   
6853  Для дома и дачи|Мебель и интерьер|Текстиль и к...   

                                          true_category  preds_ok  
8770   Для дома и дачи|Мебель и интерьер|Шкафы и комоды     False  
7744            Хобби и отдых|Коллекционирование|Другое     False  
6853  Личные вещи|Одежда, обувь, аксессуары|Женская ...     False

Видим, что в первых двух строчках ошибки, скорее всего, произошли по причине опечаток. Нужно подумать над тем, как повысить устойчивость модели к опечаткам и ошибкам. Этим вам предстоит заняться в домашнем задании.

На этом анализ ошибок, конечно же, заканчивать не стоит — но мы не будем продолжать это делать здесь, а оставим в виде упражнения.

## Word2vec

Импортируем gensim – библиотеку, с помощью которой будем обучать модель w2v. Также создадим понадобятся вспомогательные классы, которые будут логировать ошибку после каждой итерации обучения.

In [ ]:
import re
import gensim
from gensim.models.word2vec import Word2Vec
from gensim.models.fasttext import FastText
from gensim.models.callbacks import CallbackAny2Vec

class LossLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        if self.epoch == 0:
            print('Loss after epoch {}: {}'.format(self.epoch, loss))
        else:
            print('Loss after epoch {}: {}'.format(self.epoch, loss - self.loss_previous_step))
        self.epoch += 1
        self.loss_previous_step = loss

class EpochLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0

def on_epoch_end(self, model):
    print(f'Epoch {self.epoch}')
    self.epoch += 1

print(gensim.__version__)

4.0.1


/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/gensim/similarities/__init__.py:15: UserWarning: The gensim.similarities.levenshtein submodule is disabled, because the optional Levenshtein package <https://pypi.org/project/python-Levenshtein/> is unavailable. Install Levenhstein (e.g. `pip install python-Levenshtein`) to suppress this warning.
  warnings.warn(msg)


Прочитаем данные и выведем первые несколько строк:

In [ ]:
w2v_train_data = pd.read_csv('unlabeled_data.csv',  index_col=0)
print(w2v_train_data.shape)
w2v_train_data.head()

(200000, 2)


item_id                                              title
30000   377939                            1 рубль царские 10 штук
30001   384779  Напольный насос giyo высокого давления, с мано...
30002   276889                                Sony Xperia m2 Dual
30003   155718   Zoom B2.1U+ AC Процессор эффектов для бас-гитары
30004   260705                    Люстра Favourite petite 1574-7P

Данные для обучения word2vec — это просто тексты или любые другие наборы последовательностей токенов, для которых можно предположить справедливость дистрибутивной гипотезы. Токенизируем тексты:

In [ ]:
WORD_PATTERN = '(?u)\\b\\w\\w+\\b'  # паттерн для токенизации возьмем из векторайзеров sklearn
reg_exp = re.compile(pattern=WORD_PATTERN)  # скомпилируем регулярное выражение
sentences = [reg_exp.findall(s.lower()) for s in w2v_train_data.title]

Выведем результат токенизации:

In [ ]:
sentences[:3]

[['рубль', 'царские', '10', 'штук'],
 ['напольный', 'насос', 'giyo', 'высокого', 'давления', 'манометр'],
 ['sony', 'xperia', 'm2', 'dual']]

Обучим w2v с параметрами по усмолчанию:

In [ ]:
w2v_model = Word2Vec()
w2v_model.build_vocab(sentences)
w2v_model.train(
   corpus_iterable=sentences,
   total_examples=w2v_model.corpus_count,
   epochs=30,
   compute_loss=True,
   callbacks=[LossLogger()]
)

Loss after epoch 0: 707412.4375
Loss after epoch 1: 452192.8125
Loss after epoch 2: 367450.75
Loss after epoch 3: 322385.125
Loss after epoch 4: 291963.375
Loss after epoch 5: 265280.75
Loss after epoch 6: 240426.5
Loss after epoch 7: 238080.5
Loss after epoch 8: 227832.75
Loss after epoch 9: 219019.25
Loss after epoch 10: 211790.0
Loss after epoch 11: 205855.75
Loss after epoch 12: 200143.5
Loss after epoch 13: 195570.75
Loss after epoch 14: 167418.25
Loss after epoch 15: 155538.5
Loss after epoch 16: 146133.0
Loss after epoch 17: 150181.5
Loss after epoch 18: 147111.5
Loss after epoch 19: 145350.0
Loss after epoch 20: 136841.5
Loss after epoch 21: 140668.5
Loss after epoch 22: 139531.0
Loss after epoch 23: 137937.0
Loss after epoch 24: 136387.5
Loss after epoch 25: 134649.0
Loss after epoch 26: 134593.5
Loss after epoch 27: 127624.0
Loss after epoch 28: 132386.0
Loss after epoch 29: 132068.0


(17444826, 22027110)

Попытаемся понять, получилось ли обучиться. Выведем векторы, близкие к вектору слова 'продам';

In [ ]:
w2v_model.wv.similar_by_word('продам')

[('продаю', 0.9121062159538269),
 ('породам', 0.6879456639289856),
 ('отдам', 0.5875111818313599),
 ('продаются', 0.568581759929657),
 ('новую', 0.48947644233703613),
 ('подам', 0.4769991934299469),
 ('почти', 0.47238969802856445),
 ('продается', 0.4529915750026703),
 ('продаётся', 0.44807854294776917),
 ('торг', 0.4432298541069031)]

Посмотрим на ещё одно слово. Интересно, что можно поискать слова близкие к слову 'айфон' и тем самым получить написание этого слова на английском:

In [ ]:
w2v_model.wv.similar_by_word('айфон')

[('iphone', 0.8009417653083801),
 ('iphone5', 0.7927708625793457),
 ('iphon', 0.7616462707519531),
 ('iphone6', 0.738697350025177),
 ('5c', 0.733371913433075),
 ('iphone4', 0.7048424482345581),
 ('5с', 0.704780638217926),
 ('айфон4', 0.6993491053581238),
 ('ipone', 0.6901272535324097),
 ('айфонов', 0.6486436128616333)]

Напишем класс, которые позволит там получать векторы для заголовков, как сумму векторов входящих в него слов:

In [ ]:
class Word2VecTransformer:

    def __init__(self, w2v_model, word_pattern):

        self.w2v_model = w2v_model
        self.word_pattern = word_pattern
        self.re = re.compile(pattern=self.word_pattern)

    def fit(self, X):
        return self

    def transform(self, X):

        X_transformed = np.zeros((len(X), self.w2v_model.wv.vector_size))
        for i, title in enumerate(X):

            title_vector = np.zeros((self.w2v_model.wv.vector_size,))
            tokens = self.re.findall(title.lower())
            for token in tokens:
                if token in self.w2v_model.wv.key_to_index:
                    title_vector += self.w2v_model.wv.get_vector(token)

            X_transformed[i] = title_vector

        return X_transformed

Трансформируем с помощью него обучающую выборку и обучим логистическую регрессию:

In [ ]:
w2v_transformer = Word2VecTransformer(w2v_model=w2v_model, word_pattern=WORD_PATTERN)

train_w2v = w2v_transformer.transform(train_titles.values)
val_w2v = w2v_transformer.transform(val_titles.values)

model_on_w2v = LogisticRegression(solver='liblinear')
model_on_w2v.fit(train_w2v, y_train)

LogisticRegression(solver='liblinear')

Оценим качество классификатора:

In [ ]:
preds = model_on_w2v.predict(val_w2v)
accuracy_score(preds, y_val)

0.7484

Варианты улучшений:

* Word2vec не может получить вектора слов, которых не было при обучении модели. Однако есть модель, очень похожая на word2vec, но использующая при обучении буквенные n-гаммы. Эта модель называется Fasttext — в библиотеке gensim есть её реализация;
* Метод агрегации, который мы использовали, очень грубый: можно попробовать не суммировать, а считать среднее или суммировать с весами;
* Можно попробовать подобрать более оптимальные гиперпараметры w2v: использовать skipgram вместо CBOW, учить с другим размером окна, использовать другой размер вектора, изменить соотношение позитивных и негативных примеров в negative sampling или использовать иерархический софтмакс.